### PR TITLE
rust: make ExecutionContext optional in EvmcVm.execute()

### LINKED ISSUE #318: Rust bindings are incompatible with EIP-2003
[EIP-2003](https://github.com/ethereum/EIPs/pull/2003) states that `code` and (host) `context` is not mandatory for precompiles and can be null, but the Rust bindings assert for the presence of both.

### PR DESCRIPTION
**Pull request by axic**
Fixes #318.

### CODE REVIEW

File: bindings/rust/evmc-declare-tests/src/lib.rs [hunk 1]
```diff
@@ -16,7 +17,12 @@ impl EvmcVm for FooVM {
         FooVM {}
     }
 
-    fn execute(&self, _code: &[u8], _context: &ExecutionContext) -> ExecutionResult {
+    fn execute(
+        &self,
+        _code: &[u8],
```
Comment (jakelang) [inline bindings/rust/evmc-declare-tests/src/lib.rs]: I figured out recently that the identifiers for the struct members and the arguments won't conflict, therefore we don't need to prefix with an underscore.

File: examples/example-rust-vm/src/lib.rs [hunk 2]
```diff
@@ -14,8 +14,13 @@ impl EvmcVm for ExampleRustVM {
         ExampleRustVM {}
     }
 
-    fn execute(&self, _code: &[u8], context: &ExecutionContext) -> ExecutionResult {
-        let is_create = context.get_message().kind() == evmc_sys::evmc_call_kind::EVMC_CREATE;
+    fn execute(
+        &self,
+        _code: &[u8],
```
Comment (jakelang) [inline examples/example-rust-vm/src/lib.rs]: Let's get rid of this underscore to make the argument formatting consistent.

File: bindings/rust/evmc-vm/src/container.rs [hunk 3]
```diff
@@ -49,14 +49,19 @@ where
...
 
     struct TestVm {}
     impl EvmcVm for TestVm {
         fn init() -> Self {
             TestVm {}
         }
-        fn execute(&self, _code: &[u8], _context: &ExecutionContext) -> ExecutionResult {
+        fn execute(
+            &self,
+            _code: &[u8],
```
Comment (jakelang) [inline bindings/rust/evmc-vm/src/container.rs]: Drop the underscores here as well

File: bindings/rust/evmc-declare-tests/src/lib.rs (hunk 1 above)
Comment (axic) [reply bindings/rust/evmc-declare-tests/src/lib.rs]: These are in examples and are prefixed for making clippy happy since they are not used.

**Comment by axic** [comment]
> I feel like this would be cleaner if ExecutionContext would make its context member optional instead

The problem with that is every single feature of `ExecutionContext` depends on that field being present.

**Comment by chfast** [comment]
Can this be merged?

**Comment by axic** [comment]
@jakelang rebased, but not sure I like it.

@chfast we can also consider precompiles support for 6.4.

**Pull request merged**